test/src/test/java/hudson/model/QueueTest.java

```diff
@@ -152,6 +169,47 @@
         testProject.delete();
         q.load();
         assertEquals(0,q.getItems().length);
+
+        // The counter state should be maintained.
+        assertEquals(1, Queue.WaitingItem.getCurrentCounterValue());
+    }
+
+    /**
+     * Make sure the queue can be reconstructed from a List queue.xml.
+     * Prior to the Queue.State class, the Queue items were just persisted as a List.
+     */
+    @Test
+    public void recover_from_leegacy_list() throws Exception {
+        Queue q = r.jenkins.getQueue();
+
+        // prevent execution to push stuff into the queue
+        r.jenkins.setNumExecutors(0);
+        r.jenkins.setNodes(r.jenkins.getNodes());
+
+        FreeStyleProject testProject = r.createFreeStyleProject("test");
+        testProject.scheduleBuild(new UserIdCause());
+
+        // Persist the List only (as before Queue.State)
+        XmlFile xmlFile = new XmlFile(Queue.XSTREAM, queueFile);
+        xmlFile.write(new ArrayList(Arrays.asList(q.getItems())));
```

Comment:
More convincing to do this as a `@LocalData` test packaging an actual old `queue.xml` as determined experimentally.